web/src/lib/pagination.ts

```diff
@@ -1,3 +1,74 @@
+const DEFAULT_PAGE_SIZE = 10;
+
+export type Pagination = {
+	currentResultsPage: number;
+	resultsPerPage: number;
+	resultsToFetchStart: number;
+	resultsToFetchEnd: number;
+	resultShowingFirst: number;
+	resultShowingLast: number;
+	resultsCount: number;
+	pagesDisplay: Array<string>
+	pagesCount: number;
+	prevPageValid: boolean;
+	prevPageUrl: string;
+	nextPageValid: number | null;
+	nextPageUrl: string;
+};
+
+export const getPagePagination = (url: URL, resultsCount: number, resultsPerPage = DEFAULT_PAGE_SIZE ): Pagination => {
+
+	// get the base URL
+	const baseURL = url.origin + url.pathname;
+
+	// from the given URL, grab which page of results is currently selected (default 1)
+	const currentResultsPage = parseInt(url.searchParams.get('page')) || 1;
+
+	// define the start / stop range of rows we want to retreive from the DB for the page
+	const resultsToFetchStart = (currentResultsPage - 1) * resultsPerPage;
+	const resultsToFetchEnd = resultsToFetchStart + resultsPerPage;
+
+	// of the results, the first / last displayed on the page
+	const resultShowingFirst = (currentResultsPage - 1) * resultsPerPage + 1;
+	const resultShowingLast = Math.min(
+		currentResultsPage * resultsPerPage,
+		resultsCount
+	);
+
+	// the total number of pages
+	const pagesCount = Math.ceil(resultsCount / resultsPerPage)
+
+	// determine, based on the current page, if the prev / next pages are valid
+	const prevPageValid = (currentResultsPage > 1);
+	const nextPageValid = (currentResultsPage < pagesCount);
+
+	// if the prev / next page are valid, store what page they should correspond to
+	const prevPageUrl = prevPageValid ? `${baseURL}?page=${currentResultsPage - 1 }` : "javascript:void(0)";
```

Comment:
Found this `"javascript:void(0)"` solution on: https://stackoverflow.com/questions/2091168/how-to-disable-a-link-using-only-css

Seems kind of hacky, but works well because `class:cursor-not-allowed={!pagination.nextPageValid}` will still allow you to click on the link unless the href has this specific value

Reply:
Too hacky. No need to worry about downstream UI logic from this function. I would either set to `null` or to the current `url`

Reply:
Also, this logic `${baseURL}?page=${currentResultsPage - 1 }` will not work if the original `url` has other query parameters or fragments, like `/jobs#tab-2?page=1&active=true`. Use [URL search params](https://developer.mozilla.org/en-US/docs/Web/API/URL/searchParams) to set the page parameter

Reply:
The null works, not sure how I didn't come across that the first time.

I update the `prevPageUrl` assignment, but let me know if that code looks kinda jank, couldn't figure out a cleaner way to do it.